subsys/net/ip/dhcpv4.c

```diff
@@ -1069,11 +1072,40 @@ static enum net_verdict net_dhcpv4_input(struct net_conn *conn,
 	return NET_DROP;
 }
 
+static void iface_handler(struct net_mgmt_event_callback *cb,
+			  u32_t mgmt_event, struct net_if *iface)
+{
+	if (iface->config.dhcpv4.state == NET_DHCPV4_INIT) {
+		return;
+	}
+
+	if (mgmt_event == NET_EVENT_IF_DOWN) {
+		NET_DBG("Interface %p going down", iface);
+
+		iface->config.dhcpv4.attempts = 0;
+		iface->config.dhcpv4.state = NET_DHCPV4_RENEWING;
+		NET_DBG("enter state=%s",
+			net_dhcpv4_state_name(iface->config.dhcpv4.state));
+	} else if (mgmt_event == NET_EVENT_IF_UP) {
+		NET_DBG("Interface %p coming up", iface);
+
+		if (iface->config.dhcpv4.state == NET_DHCPV4_RENEWING) {
+			send_request(iface);
+		}
+	} else {
+		return;
+	}
+}
+
 void net_dhcpv4_start(struct net_if *iface)
 {
 	u32_t timeout;
 	u32_t entropy;
 
+	net_mgmt_init_event_callback(&mgmt4_cb, iface_handler,
+				     NET_EVENT_IF_DOWN | NET_EVENT_IF_UP);
+	net_mgmt_add_event_callback(&mgmt4_cb);
+
```

Comment:
Thing is, it won't work if the cable is not plugged in at start up but plugged later. So before k_delayed_work_submit() it should verify that net_if is up.

and maybe it should register the event handler after initialization.